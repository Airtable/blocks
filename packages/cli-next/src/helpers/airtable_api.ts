import fetch from 'node-fetch';
import {spawnUnexpectedError, spawnUserError} from './error_utils';
import {FetchApi, Response, FetchInit} from './fetch_api';
import {S3Api, S3SignedUploadInfo} from './s3_api';

/**
 * Enumeration of some error types server may return.
 */
/* eslint-disable airtable/enum-style */
enum ErrorResponseCode {
    RESOURCE_NOT_FOUND = 'RESOURCE_NOT_FOUND',
    UNSUPPORTED_BLOCKS_CLI_VERSION = 'UNSUPPORTED_BLOCKS_CLI_VERSION',
}
/* eslint-enable airtable/enum-style */

export enum AirtableApiErrorName {
    AIRTABLE_API_ERROR_STATUS_AND_MESSAGES = 'airtableApiErrorStatusAndMessages',
    AIRTABLE_API_KEY_MALFORMED = 'airtableApiKeyMalformed',
    AIRTABLE_API_KEY_NAME_INVALID = 'airtableApiKeyNameInvalid',
    AIRTABLE_API_MULTIPLE_ERRORS = 'airtableApiMultipleErrors',
    AIRTABLE_API_WITH_INVALID_API_KEY = 'airtableApiWithInvalidApiKey',
    AIRTABLE_API_BLOCK_NOT_FOUND = 'airtableApiBlockNotFound',
    AIRTABLE_API_UNSUPPORTED_BLOCKS_CLI_VERSION = 'airtableApiUnsupportedBlocksCliVersion',
    AIRTABLE_API_UNEXPECTED_ERROR = 'airtableApiUnexpectedError',
}

export interface AirtableApiErrorStatusAndMessages {
    type: AirtableApiErrorName.AIRTABLE_API_ERROR_STATUS_AND_MESSAGES;
    status: number;
    errors: Exclude<ErrorResponseJson['errors'], undefined>;
}

export interface AirtableApiErrorKeyMalformed {
    type: AirtableApiErrorName.AIRTABLE_API_KEY_MALFORMED;
}

export interface AirtableApiErrorKeyNameInvalid {
    type: AirtableApiErrorName.AIRTABLE_API_KEY_NAME_INVALID;
    name: string;
}

export interface AirtableApiErrorMultiple {
    type: AirtableApiErrorName.AIRTABLE_API_MULTIPLE_ERRORS;
    errors: AirtableApiErrorInfo[];
}

export interface AirtableApiErrorInvalidApiKey {
    type: AirtableApiErrorName.AIRTABLE_API_WITH_INVALID_API_KEY;
}

export interface AirtableApiErrorBaseNotFound {
    type: AirtableApiErrorName.AIRTABLE_API_BLOCK_NOT_FOUND;
}

export interface AirtableApiErrorUnsupportedBlocksCliVersion {
    type: AirtableApiErrorName.AIRTABLE_API_UNSUPPORTED_BLOCKS_CLI_VERSION;
    serverMessage: string;
}

export interface AirtableApiErrorUnexpected {
    type: AirtableApiErrorName.AIRTABLE_API_UNEXPECTED_ERROR;
    serverMessage: string;
}

export type AirtableApiErrorInfo =
    | AirtableApiErrorStatusAndMessages
    | AirtableApiErrorKeyMalformed
    | AirtableApiErrorKeyNameInvalid
    | AirtableApiErrorMultiple
    | AirtableApiErrorInvalidApiKey
    | AirtableApiErrorBaseNotFound
    | AirtableApiErrorUnsupportedBlocksCliVersion
    | AirtableApiErrorUnexpected;

interface ErrorResponseJson {
    error?: {
        code?: ErrorResponseCode | string;
        message: string;
    };
    errors?: {
        code?: ErrorResponseCode | string;
        message: string;
    }[];
}

interface ErrorResponseError {
    status: number;
    error: {
        message: string;
        code?: string;
    };
}

export interface AirtableApiOptions {
    blockId: string;
    apiKey: string;
    userAgent: string;
    apiBaseUrl: string;
}

export interface CreateBuildOptions {
    s3: S3Api;
    frontendBundle: Buffer;
    backendBundle: Buffer | null;
}

export interface CreateReleaseOptions {
    buildId: string;
    deployId?: string;
    developerComment?: string;
}

export interface CodeUploadOptions {
    codeUploadId: string;
    status: 'uploaded' | 'failed';
}

export interface CodeUploadResponse {
    codeUploadId: string;
    presignedUploadUrl: string;
}

export interface FinalizeCodeUploadResponse {
    message: string;
}

export interface UploadSubmissionOptions {
    archiveBuffer: Buffer;
}

export interface CreateBuildResponseJson {
    buildId: string;
    frontendBundleUploadUrl: string;
    backendDeploymentPackageUploadUrl: string | null;
    frontendBundleS3UploadInfo: S3SignedUploadInfo;
    backendDeploymentPackageS3UploadInfo: S3SignedUploadInfo | null;
}

function createAirtableApiError({
    status,
    error: {message, code},
}: ErrorResponseError): AirtableApiErrorInfo {
    if (status === 401) {
        return {
            type: AirtableApiErrorName.AIRTABLE_API_WITH_INVALID_API_KEY,
        };
    } else if (status === 422 && code === ErrorResponseCode.RESOURCE_NOT_FOUND) {
        return {
            type: AirtableApiErrorName.AIRTABLE_API_BLOCK_NOT_FOUND,
        };
    } else {
        return {
            type: AirtableApiErrorName.AIRTABLE_API_UNEXPECTED_ERROR,
            serverMessage: message,
        };
    }
}

export abstract class AirtableApi extends FetchApi {
    protected blockId: string;
    private apiKey: string;
    private userAgent: string;
    private apiBaseUrl: string;

    constructor({blockId, apiKey, userAgent, apiBaseUrl}: AirtableApiOptions) {
        super();
        this.blockId = blockId;
        this.apiKey = apiKey;
        this.userAgent = userAgent;
        this.apiBaseUrl = apiBaseUrl;
    }

    airtableFetchInit({url, ...init}: FetchInit): FetchInit {
        return {
            url: new URL(url, this.apiBaseUrl).href,
            ...init,
            method: 'post',
            headers: {
                Authorization: `Bearer ${this.apiKey}`,
                'User-Agent': this.userAgent,
                'Content-Type': 'application/json',
            },
        };
    }

    abstract async createBuildAsync(options: CreateBuildOptions): Promise<CreateBuildResponseJson>;

    abstract async createReleaseAsync(options: CreateReleaseOptions): Promise<void>;

    async uploadSubmissionAsync({archiveBuffer}: UploadSubmissionOptions): Promise<string> {
        const {presignedUploadUrl, codeUploadId} = await this._blockCreateCodeUploadAsync();

        const response = await fetch(presignedUploadUrl, {method: 'put', body: archiveBuffer});

        const didUpload = response.status === 200;
        const status = didUpload ? 'uploaded' : 'failed';
        const {message} = await this._blockFinalizeCodeUploadAsync({
            codeUploadId,
            status,
        });

        if (!didUpload) {
            throw spawnUnexpectedError(message);
        }

        return message;
    }

    protected abstract async _blockCreateCodeUploadAsync(): Promise<CodeUploadResponse>;

    protected abstract async _blockFinalizeCodeUploadAsync(
        options: CodeUploadOptions,
    ): Promise<FinalizeCodeUploadResponse>;

    protected async _invariantOkResponseAsync(init: FetchInit, response: Response) {
        const {status} = response;
        if (status === 200) {
            return;
        }

        const {error, errors}: ErrorResponseJson = await response.json();
        if (Array.isArray(errors) && errors.length > 1) {
            if (errors.some(errObj => errObj && typeof errObj.message === 'string')) {
                throw spawnUserError<AirtableApiErrorInfo>({
                    type: AirtableApiErrorName.AIRTABLE_API_ERROR_STATUS_AND_MESSAGES,
                    status,
                    errors,
                });
            }
            const errorMessages = errors.map(err => createAirtableApiError({status, error: err}));
            throw spawnUserError<AirtableApiErrorInfo>({
                type: AirtableApiErrorName.AIRTABLE_API_MULTIPLE_ERRORS,
                errors: errorMessages,
            });
        } else {
            const errorToThrow = Array.isArray(errors) && errors.length === 1 ? errors[0] : error;
            if (
                errorToThrow &&
                errorToThrow.code === ErrorResponseCode.UNSUPPORTED_BLOCKS_CLI_VERSION &&
                typeof errorToThrow.message === 'string'
            ) {
                throw spawnUserError<AirtableApiErrorInfo>({
                    type: AirtableApiErrorName.AIRTABLE_API_UNSUPPORTED_BLOCKS_CLI_VERSION,
                    serverMessage: errorToThrow.message,
                });
            } else if (errorToThrow && typeof errorToThrow.message === 'string') {
                throw spawnUserError<AirtableApiErrorInfo>(
                    createAirtableApiError({status, error: errorToThrow}),
                );
            } else {
                throw spawnUnexpectedError(
                    'Request to Airtable failed with status code %s',
                    status,
                );
            }
        }
    }
}
