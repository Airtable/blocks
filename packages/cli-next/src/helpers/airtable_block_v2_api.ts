import {
    AirtableApi,
    AirtableApiOptions,
    CodeUploadOptions,
    CodeUploadResponse,
    CreateBuildOptions,
    CreateBuildResponseJson,
    CreateReleaseOptions,
    FinalizeCodeUploadResponse,
} from './airtable_api';
import {FetchInit} from './fetch_api';
import {S3SignedUploadInfo} from './s3_api';
import {invariant, spawnUserError} from './error_utils';
import {ReleaseCommandErrorInfo, ReleaseCommandErrorName} from './release_messages';

export interface AirtableBlockV2ApiBaseOptions extends AirtableApiOptions {
    blockId: string;
}

export interface AirtableBlockV2ApiReleaseOptions {
    buildId: string;
    developerComment: string;
}

export interface AirtableBlockV2ApiBuildStartResponse {
    buildId: string;
    frontendBundleS3UploadInfo: S3SignedUploadInfo;
}

export interface AirtableBlockV2ApiReleaseResponse {
    releaseId: string;
}

export class AirtableBlockV2Api extends AirtableApi {
    airtableFetchInit({url, ...init}: FetchInit): FetchInit {
        return super.airtableFetchInit({
            url: `/v2/blocksV2/${this.blockId}${url}`,
            ...init,
        });
    }

    async _blockBuildStartAsync(): Promise<CreateBuildResponseJson> {
        return await this.fetchJsonAsync(this.airtableFetchInit({url: '/builds/start'}));
    }

    async blockCreateReleaseAsync({
        buildId,
        developerComment,
    }: AirtableBlockV2ApiReleaseOptions): Promise<AirtableBlockV2ApiReleaseResponse> {
        return await this.fetchJsonAsync(
            this.airtableFetchInit({
                url: '/releases/create',
                body: {buildId, developerComment},
            }),
        );
    }

    async _blockCreateCodeUploadAsync(): Promise<CodeUploadResponse> {
        return await this.fetchJsonAsync(
            this.airtableFetchInit({
                url: '/codeUpload/create',
                body: {isV2Block: true},
            }),
        );
    }

    async _blockFinalizeCodeUploadAsync({
        codeUploadId,
        status,
    }: CodeUploadOptions): Promise<FinalizeCodeUploadResponse> {
        return await this.fetchJsonAsync(
            this.airtableFetchInit({
                url: '/codeUpload/finalize',
                body: {
                    codeUploadId,
                    status,
                },
            }),
        );
    }

    async createBuildAsync({
        s3,
        frontendBundle,
        backendBundle,
    }: CreateBuildOptions): Promise<CreateBuildResponseJson> {
        if (backendBundle) {
            throw spawnUserError<ReleaseCommandErrorInfo>({
                type: ReleaseCommandErrorName.RELEASE_COMMAND_BLOCK2_BACKEND_UNSUPPORTED,
            });
        }

        const build = await this._blockBuildStartAsync();

        await s3.uploadBundleAsync({
            fileBuffer: frontendBundle,
            uploadInfo: build.frontendBundleS3UploadInfo,
        });

        return build;
    }

    async createReleaseAsync({
        buildId,
        deployId,
        developerComment,
    }: CreateReleaseOptions): Promise<void> {
        if (deployId) {
            throw spawnUserError<ReleaseCommandErrorInfo>({
                type: ReleaseCommandErrorName.RELEASE_COMMAND_BLOCK2_BACKEND_UNSUPPORTED,
            });
        }
        invariant(developerComment, 'developerComment is required for releasing a v2 block');
        await this.blockCreateReleaseAsync({
            buildId,
            developerComment: developerComment,
        });
    }
}
