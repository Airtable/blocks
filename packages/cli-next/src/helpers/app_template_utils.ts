
import {Octokit} from '@octokit/rest';
import tar from 'tar';
import fetch from 'node-fetch';
import {System} from './system';
import {invariant} from './error_utils';
import {mkdirpAsync} from './system_extra';

export async function getGithubTemplateTarballUrlAsync(templateUrl: string) {
    const octokit = new Octokit({});
    const pathComponents = new URL(templateUrl).pathname.split('/');
    const [, owner, repo] = pathComponents;

    try {
        const {url} = await octokit.repos.downloadTarballArchive({
            owner,
            repo,
            ref: '',
        });
        return url;
    } catch (e) {
        if (e.status === 404) {
            e.message = `Template ${templateUrl} is not a public repo that could be found on Github`;
        }
        throw e;
    }
}

export async function downloadTarballAsync(
    sys: System,
    tarballUrl: string,
    destinationPath: string,
): Promise<void> {
    await mkdirpAsync(sys, sys.path.dirname(destinationPath));

    const response = await fetch(tarballUrl);
    const buffer = await response.buffer();

    await sys.fs.writeFileAsync(destinationPath, buffer);
}

export async function extractTarballAsync(
    sys: System,
    tarballPath: string,
    destinationDir: string,
): Promise<string> {
    await sys.fs.mkdirAsync(destinationDir, {recursive: true});
    await tar.extract({file: tarballPath, cwd: destinationDir});

    const entries = await sys.fs.readdirAsync(destinationDir, {
        withFileTypes: true,
        encoding: 'utf8',
    });
    const templateDir = entries.find(dirent => dirent.isDirectory());
    invariant(templateDir !== undefined, 'template does not contain a directory');

    return sys.path.join(destinationDir, templateDir.name);
}
