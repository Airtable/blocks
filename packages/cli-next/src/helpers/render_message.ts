import {UserError} from './error_utils';

export type SelectMessage<T extends {type: string}, Name extends T['type']> = T extends {type: Name}
    ? T
    : never;

export type MessageCalls<MessageInfo extends {type: string}, MessageUtil = void> = {
    [key in MessageInfo['type']]: (
        this: Messages<MessageInfo, MessageUtil>,
        info: SelectMessage<MessageInfo, key>,
    ) => string;
};

export type Messages<MessageInfo extends {type: string}, MessageUtil = void> = {
    util: MessageUtil;
    renderMessage(message: MessageInfo): string;
    renderError(err: UserError<MessageInfo>): string;
    supportsMessage<Info extends MessageInfo = MessageInfo>(message: unknown): message is Info;
    supportsError<Info extends MessageInfo = MessageInfo>(err: Error): err is UserError<Info>;
} & MessageCalls<MessageInfo>;

export interface MessageConstructor<MessageInfo extends {type: string}, MessageUtil = void> {
    new (util: MessageUtil): Messages<MessageInfo>;
}

export class RenderMessage<MessageInfo extends {type: string}, MessageUtil = void> {
    util: MessageUtil;

    constructor(util: MessageUtil) {
        this.util = util;
    }

    renderMessage(message: MessageInfo): string {
        return (this as any)[message.type](message);
    }
    renderError(err: UserError<MessageInfo>): string {
        return this.renderMessage(err.__userInfo);
    }
    supportsMessage(message: unknown): message is MessageInfo {
        return (
            typeof message === 'object' &&
            message !== null &&
            'type' in message &&
            (message as MessageInfo).type in this
        );
    }
    supportsError(err: Error & {__userInfo?: unknown}): err is UserError<MessageInfo> {
        return this.supportsMessage(err.__userInfo);
    }
    static extend<ExtendMessageInfo extends {type: string}, ExtendMessageUtil = void>(
        methods: MessageCalls<ExtendMessageInfo, ExtendMessageUtil>,
    ): MessageConstructor<ExtendMessageInfo, ExtendMessageUtil> {
        const ExtendedFormatMessage = class ExtendedFormatMessage extends RenderMessage<
            ExtendMessageInfo,
            ExtendMessageUtil
        > {};
        Object.assign(ExtendedFormatMessage.prototype, methods);
        return ExtendedFormatMessage as any;
    }
}
