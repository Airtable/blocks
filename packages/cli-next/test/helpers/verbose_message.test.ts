import chalk from 'chalk';

import {test, expect} from '../mocks/test';

import {
    AirtableApiErrorName,
    AppConfigErrorName,
    BlockIdentifierErrorName,
    BuildErrorName,
    DevelopmentRunFrameMessageName,
    FindPortErrorName,
    InitCommandErrorName,
    InitCommandMessageName,
    MessageInfo,
    MessageName,
    ReleaseCommandErrorName,
    ReleaseCommandMessageName,
    RemoteCommandErrorName,
    RemoteCommandMessageName,
    RemoteConfigErrorName,
    RunCommandMessageName,
    S3ApiErrorName,
    SubmitCommandErrorName,
    SubmitCommandMessageName,
    SystemApiKeyErrorName,
    SystemConfigErrorName,
    SystemExtraErrorName,
    UserConfigErrorName,
    VerboseMessage,
} from '../../src/helpers/verbose_message';
import {SelectMessage} from '../../src/helpers/render_message';
import {
    BLOCK_CONFIG_DIR_NAME,
    BLOCK_FILE_NAME,
    INIT_DEFAULT_TEMPLATE_URL,
    REMOTE_JSON_BASE_FILE_PATH,
    USER_CONFIG_FILE_NAME,
} from '../../src/settings';

describe('verbose_message', function() {
    test.it('at least one test per message type', () => {
        const tests = testMessages();
        const missingTest = Object.values(MessageName).filter(
            type => !tests[type] || tests[type].length === 0,
        );
        expect(missingTest).to.deep.equal([]);
    });

    for (const [type, typeSuite] of Object.entries(testMessages())) {
        for (const info of typeSuite) {
            test.it(`formats ${type}`, () => {
                expect(
                    new VerboseMessage({chalk: new chalk.Instance({level: 3})}).renderMessage({
                        type: type as any,
                        ...info,
                    }),
                ).to.matchSnapshot();
            });
        }
    }
});

function testMessages(): {
    [key in MessageName]: Omit<SelectMessage<MessageInfo, key>, 'type'>[];
} {
    return {
        [AirtableApiErrorName.AIRTABLE_API_BLOCK_NOT_FOUND]: [{}],
        [AirtableApiErrorName.AIRTABLE_API_ERROR_STATUS_AND_MESSAGES]: [
            {status: 400, errors: []},
            {
                status: 400,
                errors: [{code: 'SOMETHING_WRONG', message: 'Something went wrong.'}],
            },
        ],
        [AirtableApiErrorName.AIRTABLE_API_KEY_MALFORMED]: [{}],
        [AirtableApiErrorName.AIRTABLE_API_KEY_NAME_INVALID]: [
            {name: 'a taxi cab covered in waffles'},
        ],
        [AirtableApiErrorName.AIRTABLE_API_MULTIPLE_ERRORS]: [
            {errors: []},
            {
                errors: [
                    {type: AirtableApiErrorName.AIRTABLE_API_WITH_INVALID_API_KEY},
                    {
                        type: AirtableApiErrorName.AIRTABLE_API_UNEXPECTED_ERROR,
                        serverMessage: 'Something went wrong.',
                    },
                ],
            },
        ],
        [AirtableApiErrorName.AIRTABLE_API_UNSUPPORTED_BLOCKS_CLI_VERSION]: [
            {serverMessage: 'Please update @airtable/blocks-cli.'},
        ],
        [AirtableApiErrorName.AIRTABLE_API_WITH_INVALID_API_KEY]: [{}],
        [AirtableApiErrorName.AIRTABLE_API_UNEXPECTED_ERROR]: [
            {serverMessage: 'Something went wrong.'},
        ],

        [AppConfigErrorName.APP_CONFIG_IS_NOT_VALID]: [
            {message: 'should be a non-null object'},
            {file: `../${BLOCK_FILE_NAME}`, message: 'should be a non-null object'},
        ],

        [BlockIdentifierErrorName.BLOCK_IDENTIFIER_INVALID_BASE_ID]: [{}],
        [BlockIdentifierErrorName.BLOCK_IDENTIFIER_INVALID_BLOCK_ID]: [{}],
        [BlockIdentifierErrorName.BLOCK_IDENTIFIER_INVALID_FORMAT]: [{}],

        [BuildErrorName.BUILD_APP_CONFIG_MODIFIED]: [{}],
        [BuildErrorName.BUILD_BLOCK_DIRECTORY_NOT_FOUND]: [{}],

        [DevelopmentRunFrameMessageName.DEVELOPMENT_RUN_FRAME_NEW_BLOCK_INSTALLATION]: [{}],
        [DevelopmentRunFrameMessageName.DEVELOPMENT_RUN_FRAME_ORIGINAL_BASE_ONLY]: [{}],
        [DevelopmentRunFrameMessageName.DEVELOPMENT_RUN_FRAME_ORIGINAL_BLOCK_ONLY]: [{}],

        [FindPortErrorName.FIND_PORT_ASYNC_PORT_IS_NOT_NUMBER]: [{port: 'asdf'}],

        [InitCommandErrorName.INIT_COMMAND_DIRECTORY_EXISTS]: [
            {
                blockDirPath: 'my-app',
            },
        ],
        [InitCommandErrorName.INIT_COMMAND_INSTALLED_SDK_NO_VERSION]: [{}],
        [InitCommandErrorName.INIT_COMMAND_TEMPLATE_MISSING]: [
            {
                template: INIT_DEFAULT_TEMPLATE_URL,
            },
        ],
        [InitCommandErrorName.INIT_COMMAND_TEMPLATE_NO_BLOCK_JSON]: [
            {
                template: INIT_DEFAULT_TEMPLATE_URL,
            },
        ],
        [InitCommandErrorName.INIT_COMMAND_UNKNOWN_ERROR]: [{}],
        [InitCommandMessageName.INIT_COMMAND_READY]: [
            {
                blockDirPath: 'my-app',
                platform: 'win32',
            },
            {
                blockDirPath: 'my-app',
                platform: 'darwin',
            },
            {
                blockDirPath: 'my-app',
                platform: 'linux',
            },
        ],

        [ReleaseCommandErrorName.RELEASE_COMMAND_BLOCK1_COMMENT_UNSUPPORTED]: [{}],
        [ReleaseCommandErrorName.RELEASE_COMMAND_BLOCK2_BACKEND_UNSUPPORTED]: [{}],
        [ReleaseCommandMessageName.RELEASE_COMMAND_DEVELOPER_COMMENT_PROMPT]: [{}],

        [RemoteCommandErrorName.REMOTE_COMMAND_CONFIG_EXISTS]: [{remoteName: 'newremote'}],
        [RemoteCommandErrorName.REMOTE_COMMAND_CONFIG_MISSING]: [{remoteName: 'oldremote'}],
        [RemoteCommandErrorName.REMOTE_COMMAND_NO_CONFIGS]: [{}],
        [RemoteCommandMessageName.REMOTE_COMMAND_ADDED_NEW]: [
            {remoteFile: '.block/newremote.remote.json'},
            {remoteFile: '../.block/newremote.remote.json'},
        ],
        [RemoteCommandMessageName.REMOTE_COMMAND_BETA_WARNING]: [{}],
        [RemoteCommandMessageName.REMOTE_COMMAND_REMOVED_EXISTING]: [
            {remoteFile: '.block/oldremote.remote.json'},
            {remoteFile: '../.block/oldremote.remote.json'},
        ],

        [RemoteConfigErrorName.REMOTE_CONFIG_IS_NOT_VALID]: [
            {message: 'should be a non-null object'},
            {
                file: `../${BLOCK_CONFIG_DIR_NAME}/${REMOTE_JSON_BASE_FILE_PATH}`,
                message: 'should be a non-null object',
            },
        ],

        [RunCommandMessageName.RUN_COMMAND_INSTALLING_LOCAL_SDK]: [
            {
                sdkPath: 'path',
            },
        ],
        [S3ApiErrorName.S3_API_BUNDLE_TOO_LARGE]: [{}],
        [S3ApiErrorName.S3_API_FAILED]: [{}],

        [SubmitCommandErrorName.SUBMIT_COMMAND_WINDOWS_MULTIPLE_DISKS]: [{}],
        [SubmitCommandMessageName.SUBMIT_COMMAND_PACKAGED_CONTINUE_PROMPT]: [{}],
        [SubmitCommandMessageName.SUBMIT_COMMAND_STOP_AFTER_PACKAGING]: [{}],

        [SystemApiKeyErrorName.SYSTEM_API_KEY_NOT_FOUND]: [{}],

        [SystemConfigErrorName.SYSTEM_CONFIG_INVALID_REMOTE_NAME]: [{name: '@remotename'}],
        [SystemConfigErrorName.SYSTEM_CONFIG_APP_DIRECTORY_NOT_FOUND]: [{}],

        [SystemExtraErrorName.SYSTEM_EXTRA_DIR_WITH_FILE_NOT_FOUND]: [{file: 'block.json'}],

        [UserConfigErrorName.USER_CONFIG_IS_NOT_VALID]: [
            {message: 'should be a non-null object'},
            {
                file: `../../../.config/${USER_CONFIG_FILE_NAME}`,
                message: 'should be a non-null object',
            },
        ],
    };
}
