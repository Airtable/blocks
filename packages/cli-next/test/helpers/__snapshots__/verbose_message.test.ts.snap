// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verbose_message formats airtableApiBlockNotFound 1`] = `"‚ùå The block could not be found. Make sure your API key is correct and that you have access to the block."`;

exports[`verbose_message formats airtableApiErrorStatusAndMessages 1`] = `
"Airtable server responded with status 400:

[]"
`;

exports[`verbose_message formats airtableApiErrorStatusAndMessages 2`] = `
"Airtable server responded with status 400:

[{\\"code\\":\\"SOMETHING_WRONG\\",\\"message\\":\\"Something went wrong.\\"}]"
`;

exports[`verbose_message formats airtableApiKeyMalformed 1`] = `"‚ùå Invalid Airtable API key specified. Make sure you copied it correctly from airtable.com/account."`;

exports[`verbose_message formats airtableApiKeyNameInvalid 1`] = `"‚ùå Invalid Airtable API key name specified: \\"a taxi cab covered in waffles\\"."`;

exports[`verbose_message formats airtableApiMultipleErrors 1`] = `
"Airtable server responded with multiple errors:

"
`;

exports[`verbose_message formats airtableApiMultipleErrors 2`] = `
"Airtable server responded with multiple errors:

‚ùå Your Airtable API key is invalid. Please use [36m[1mblock set-api-key[22m[39m to update it.
‚ùå Airtable server returned an error.
Something went wrong."
`;

exports[`verbose_message formats airtableApiUnexpectedError 1`] = `
"‚ùå Airtable server returned an error.
Something went wrong."
`;

exports[`verbose_message formats airtableApiUnsupportedBlocksCliVersion 1`] = `
"‚ùå Please update @airtable/blocks-cli.

Run [36m[1mnpm i -g @airtable/blocks[22m[39m to update.
"
`;

exports[`verbose_message formats airtableApiWithInvalidApiKey 1`] = `"‚ùå Your Airtable API key is invalid. Please use [36m[1mblock set-api-key[22m[39m to update it."`;

exports[`verbose_message formats appConfigIsNotValid 1`] = `"‚ùå block.json should be a non-null object"`;

exports[`verbose_message formats appConfigIsNotValid 2`] = `"‚ùå ../block.json should be a non-null object"`;

exports[`verbose_message formats blockIdentifierInvalidBaseId 1`] = `"Block identifier's must start with \\"app\\"."`;

exports[`verbose_message formats blockIdentifierInvalidBlockId 1`] = `"Block identifier's must start with \\"blk\\" after \\"/\\"."`;

exports[`verbose_message formats blockIdentifierInvalidFormat 1`] = `"Block identifier must be in the format <baseId>/<blockId>."`;

exports[`verbose_message formats buildAppConfigModified 1`] = `"Detected changes to 'block.json' file. Please restart [36m[1mblock run[22m[39m."`;

exports[`verbose_message formats buildBlockDirectoryNotFound 1`] = `"Block directory with block.json file not found"`;

exports[`verbose_message formats developmentRunFrameNewBlockInstallation 1`] = `"Switched to a new block installation."`;

exports[`verbose_message formats developmentRunFrameOriginalBaseOnly 1`] = `"You can only run your development block in the original base where it was created."`;

exports[`verbose_message formats developmentRunFrameOriginalBlockOnly 1`] = `"blockId mismatch. Make sure you are entering development for the correct block."`;

exports[`verbose_message formats findPortAsyncPortIsNotNumber 1`] = `"‚ùå Cannot listen to port [4masdf[24m. [4masdf[24m is not a number."`;

exports[`verbose_message formats initCommandDirectoryExists 1`] = `"A directory already exists at my-app."`;

exports[`verbose_message formats initCommandInstalledSdkNoVersion 1`] = `"Installed @airtable/blocks dependency has no version."`;

exports[`verbose_message formats initCommandReady 1`] = `"‚úÖ Your block is ready! [36m[1mcd my-app[22m[39m then [36m[1mblock run[22m[39m to start developing, and [36m[1mnpm run lint[22m[39m to lint."`;

exports[`verbose_message formats initCommandReady 2`] = `"‚úÖ Your block is ready! [36m[1mcd my-app && block run[22m[39m to start developing, and [36m[1mnpm run lint[22m[39m to lint."`;

exports[`verbose_message formats initCommandReady 3`] = `"‚úÖ Your block is ready! [36m[1mcd my-app && block run[22m[39m to start developing, and [36m[1mnpm run lint[22m[39m to lint."`;

exports[`verbose_message formats initCommandTemplateMissing 1`] = `"Could not get template https://github.com/Airtable/apps-hello-world - please check you entered the name correctly."`;

exports[`verbose_message formats initCommandTemplateNoBlockJson 1`] = `"https://github.com/Airtable/apps-hello-world does not seem to be a block template."`;

exports[`verbose_message formats initCommandUnknownError 1`] = `"‚ùå Something failed! Cleaning up..."`;

exports[`verbose_message formats releaseCommandBlock1CommentUnsupported 1`] = `"‚ùå This block cannot be released with a comment. Run [36m[1mblock release[22m[39m without [36m[1m--comment[22m[39m."`;

exports[`verbose_message formats releaseCommandBlock2BackendUnsupported 1`] = `"‚ùå V2 blocks does not support backend routes."`;

exports[`verbose_message formats releaseCommandDeveloperCommentPrompt 1`] = `"Enter a comment describing the changes in this release"`;

exports[`verbose_message formats remoteCommandAddedNew 1`] = `"‚úÖ Successfully added a new remote at .block/newremote.remote.json!"`;

exports[`verbose_message formats remoteCommandAddedNew 2`] = `"‚úÖ Successfully added a new remote at ../.block/newremote.remote.json!"`;

exports[`verbose_message formats remoteCommandBetaWarning 1`] = `
"[93mNote:[39m The remotes feature is still in beta.
"
`;

exports[`verbose_message formats remoteCommandConfigExists 1`] = `
"‚ùå The [1mnewremote[22m remote already exists!
If you want to update the remote, please run [36m[1mblock remove-remote newremote[22m[39m and re-run [36m[1mblock add-remote[22m[39m!"
`;

exports[`verbose_message formats remoteCommandConfigMissing 1`] = `"‚ùå The [1moldremote[22m remote does not exist."`;

exports[`verbose_message formats remoteCommandNoConfigs 1`] = `"‚ùå This project has no remotes. Use [1mblock add-remote[22m to add one."`;

exports[`verbose_message formats remoteCommandRemovedExisting 1`] = `"‚úÖ Successfully removed the remote from .block/oldremote.remote.json!"`;

exports[`verbose_message formats remoteCommandRemovedExisting 2`] = `"‚úÖ Successfully removed the remote from ../.block/oldremote.remote.json!"`;

exports[`verbose_message formats remoteConfigIsNotValid 1`] = `"‚ùå .block/remote.json should be a non-null object"`;

exports[`verbose_message formats remoteConfigIsNotValid 2`] = `"‚ùå ../.block/remote.json should be a non-null object"`;

exports[`verbose_message formats runCommandInstallingLocalSdk 1`] = `"Installing local SDK from path..."`;

exports[`verbose_message formats s3ApiBundleTooLarge 1`] = `"‚ùå Could not upload bundle. The bundle's size is too big."`;

exports[`verbose_message formats s3ApiFailed 1`] = `"‚ùå Could not upload bundle. Failed to upload."`;

exports[`verbose_message formats submitCommandPackagedContinuePrompt 1`] = `
"
The project source code has been packaged for review by Airtable.
Are you ready to submit it for review? (Enter [36m[1my[22m[39m to continue)"
`;

exports[`verbose_message formats submitCommandStopAfterPackaging 1`] = `"Quitting. If you want to submit later run this command again."`;

exports[`verbose_message formats submitCommandWindowsMultipleDisks 1`] = `"Dependencies cannot be on multiple disks."`;

exports[`verbose_message formats systemApiKeyNotFound 1`] = `"‚ùå An airtable api key is not set. Please use [36m[1mblock set-api-key[22m[39m to set it."`;

exports[`verbose_message formats systemConfigAppDirectoryNotFound 1`] = `"‚ùå Could not find the root of an Airtable project. Please run again from an Airtable project."`;

exports[`verbose_message formats systemConfigInvalidRemoteName 1`] = `"‚ùå Incorrect characters for the name! Only alphanumeric, -, or _ characters are allowed"`;

exports[`verbose_message formats systemExtraDirWithFileNotFound 1`] = `"‚ùå Could not find directory that includes a block.json entry."`;

exports[`verbose_message formats userConfigIsNotValid 1`] = `"‚ùå ~/.config/.airtableblocksrc.json should be a non-null object"`;

exports[`verbose_message formats userConfigIsNotValid 2`] = `"‚ùå ../../../.config/.airtableblocksrc.json should be a non-null object"`;
